content/learning-paths/smartphones-and-mobile/using-neon-intrinsics-to-optimize-unity-on-android/5-the-optimizations.md

```diff
@@ -0,0 +1,275 @@
+
+```
+Line 736: wallMinXs = vld1q_f32(walls + w)
+```
+Load multiple floating point values from memory into a single vector register. The returned vector will contain _Min X_ values from 4 different walls.
+
+```
+Line 741: vcgeq_f32(wallMinXs, charMaxXs)
+```
+Floating point comparison (greater-than or equal). Compares 4 walls at once with a character’s _Max X_. Each of the four results will either be all ones (true) or all zeros (false).
+
+```
+Line 741: vorrq_u32(vcgeq_f32(wallMinXs, charMaxXs), vcgeq_f32(wallMinYs, charMaxYs))
+```
+Bitwise inclusive OR. The nested calls to _vcgeq\_f32_ are comparing the walls (Min X and Min Y) against the character’s Max X and Max Y. The four comparison results are combined with a bitwise OR. 
+
+```
+Line 741: results = vqtbl1_u8(_result of ORs_, tblindex1)
+```
+Table lookup function that selects elements from an array based on the indices provided. The result of the OR operations will be treated as an array of 8-bit values. The values from _tblindex1_ (0, 4, 8 and 12) ensure that we select the most significant bytes from each u32 OR result. So 4 character-wall comparisons are being merged into one 128-bit vector along with 4 dummy (because of the _out of range_ values in tblindex1) that will be replaced later with the 64-bit value of the next 4 wall comparisons (from _wmvn_u8_).
+
+```
+Line 751: vqtbx1_u8(results, …)
+```
+Table lookup function except that when an index is out of range, it leaves the existing data alone. This has the effect of selecting 4 bytes (using indices from _tblindex2_) from the results of the last 4 comparisons and combines it with the previous 4 results. _results_ will now contain the results of 8 wall-character comparisons.
+
+```
+Line 751: results = vmvn_u8(...)
+```
+Bitwise NOT operation. This negates each of the 8 character-wall comparisons. It is effectively the _!_ (NOT) in our intersect function; except that it is working on 8 results instead of 1.
+
+```
+Line 755: *(Unity.Burst.Intrinsics.v64*)(collisions + (c * numWalls + w - 4)) = results;
+```
+This is a vectorized form of writing out the 8 results to memory all at once. At time of writing, Burst doesn’t include the _VSTR_ Neon intrinsic.
+
+### Character-character collision detection
+This is left as an exercise for the reader. You may wish to explore the _NeonRadiusObjCollisionDetectionUnrolled_ function starting on line 561 of _CollisionCalculationScript.cs_.
+
+## Additional notes
+The sample project increases the character count over time. Because of this it was better to profile using later frames. In early frames there aren’t enough characters so the overhead of vectorization counters any gains.
+
+An early implementation used Burst Jobs. The jobs got in the way of performance timing because they are not guaranteed to execute immediately.
+
+[NoAlias] had a big impact on the performance gains (without it some optimizations produced little to no gain).
```

Comment:
worth saying [NoAlias] was explained above?

Reply:
added with link to best practice
done